Remove modulo bias (IPS-1072)

tomauth0 opened the pull request:
### Description

This change updates the id generation algorithm to remove [modulo bias](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modulo_bias) when selecting characters for ids.  

It implements a randomInteger function, which converts and samples using the [simple discard method](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) (p69).

Note that by removing the simple modulo sampling, there is a decrease in performance - the algorithm no longer runs in a fixed length of time, and overall expected processing time is increased.  
For example, for generating 100000 ids, length 16 chars and prefixed with 'cus':

modulo execution time:
264.856ms
271.453ms
275.711ms

randomInteger execution time:
2704.586ms
2682.714ms
2709.158ms

### References

Internal ticket references:

IPS-1072
ESD-4441

### Testing

Tested in node version 8, 10, 12.
Test coverage is extended to include the randomInt function.

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

tomauth0 committed:
Added randomInt backport

To remove modulo bias when selecting id characters, introduce a backport of the nodejs crypto.randomInt into the core id gen function.

Added an explicit license file for that code and referenced it from randomInt implementation.

tomauth0 committed:
cleanup test files

tomauth0 committed:
applied randomInt to IdGenerator.get

tomauth0 committed:
remove unneeded crypto.randomBytes

tomauth0 committed:
bump npm version

luuuis commented on lib/index.js:
```diff
@@ -2,7 +2,94 @@ const crypto = require('crypto');
+ This version eplaces native NumberIsSafeInteger 
+ with Number.isSafeInteger and uses locally defined rand_max values
+
+ This function is licensed under the terms of the nodejs license,
```
From a licensing POV it is best to keep code from different sources segregated; could you move this into its own file with a license header?

Even though auth0/id-generator is itself OSS I think it is helpful to keep these separate.

esarafianou replied:
Might be worth asking our Compliance team.

tomauth0 replied:
Switched the function to it's own file with an explicit license statement.  I'll reach out to compliance for guidance now

FadyMak commented on lib/index.js:
```diff
@@ -2,7 +2,94 @@ const crypto = require('crypto');
+      return n;
+    }
+  } else {
+    // Async API
```
Not a big deal since this is well-tested code that is brought over: my concern is adding code that may not be exercised in tests nor in practice since we don't expose an async API for this lib and we may deviate in the sync implementation over time.

tomauth0 replied:
I've removed the async code path from the function now, and created a set of unit tests that hopefully give some good coverage.

FadyMak reviewed (changes requested):
Nice work Tom 👏  I left a comment about the `undefined` internal errors ported over from the Node module when you get a chance to have a quick look.

For the comment about the async implementation ported over: feel free to disregard it if you believe we should keep it around.

I think it could be helpful to add some tests to cover some of the code ported over to exercise some of the "less-travelled" paths. It doesn't need to be anything too in-depth since the code is already well tested. What do you think?

tomauth0 committed:
move randomInt to new file for clearer license boundaries
change randomInt implementation to use standard errors
remove async code path from randomInt
add specific unit tests for random int - exceptional behaviour and limited boundary checks

tomauth0 committed:
missed native exception def

tomauth0 commented on lib/index.js:
```diff
@@ -2,7 +2,94 @@ const crypto = require('crypto');
+
+  const isSync = typeof callback === 'undefined';
+  if (!isSync && typeof callback !== 'function') {
+    throw new ERR_INVALID_CALLBACK(callback);
```
I've switched the implementation to the Error exceptions used in the rest of the library

tomauth0 committed:
simplify randomInteger implementation, explicitly block support for negative numbers - we only support positive ranges.

clean up decls in index.js

luuuis commented on test/randomInt.tests.js:
```diff
@@ -0,0 +1,29 @@
+const expect = require('chai').expect;
```
File should be named `randomInteger.tests.js`

luuuis commented on lib/randomInteger.js:
```diff
@@ -0,0 +1,51 @@
+const crypto = require('crypto');
+
+// Largest integer we can read from a buffer.
+const RAND_MAX = Buffer.from("ff".repeat(6), "hex").readUIntBE(0, 6);
```
This presupposes [Node >= 5.10.0](https://nodejs.org/api/buffer.html#buffer_static_method_buffer_from_string_encoding) as do the `let` changes above. It's fine by me but can you verify that this is not going to cause disruption for any libraries that consume this?

We should add the minimum version to package.json.

tomauth0 replied:
I've switched this to a static def (281474976710655) and rolled back let/const changes - these are var now for wider compatibility.  
I've tested importing the lib in v4, v10 & v15 - v4 imports correctly, but tests do not execute due to mocha engine requirements.

tomauth0 committed:
Changed RAND_MAX const in randomInteger.js to a statically defined value.
Rolled back const & let defs to var for wider node compatibility.
Bumped package version to 0.3.0

tomauth0 committed:
Hoist the randomInteger.js argument validation to before local or native implementations to enable predictable exceptions on newer versions of node with a native randomInt implementation.

tomauth0 committed:
fix over-bump :)

luuuis commented on lib/index.js:
```diff
@@ -12,56 +12,51 @@ const IdGenerator = function(prefixes){
 
   for (var i = 0; i < this.len; i++) {
-    value[i] = this.alphabet[rnd[i] % charsLength];
+    value[i] = this.alphabet[randomInteger(charsLength - 1)];
```
If `maximum` is exclusive should this not be `this.alphabet[randomInteger(charsLength)]`?

luuuis commented on lib/randomInteger.js:
```diff
@@ -0,0 +1,51 @@
+ * Generates an integer between 0 and maximum range where 0 is inclusive and maximum is
+ * exclusive.
+ * @param maximum { number } a natural number specifying the maximum random value that can be returned
+ * @returns { number } a natural number between 0 and the specified maximum
```
We should indicate that it's exclusive of maximum.

luuuis commented on lib/index.js:
```diff
@@ -12,56 +12,51 @@ const IdGenerator = function(prefixes){
-
   for (var i = 0; i < len; i++) {
-    value[i] = ALPHA_NUM[rnd[i] % charsLength];
+    value[i] = ALPHA_NUM[randomInteger(ALPHA_NUM.length - 1)];
```
If `maximum` is exclusive should this not be `ALPHA_NUM[randomInteger(ALPHA_NUM.length)]`?

tomauth0 committed:
fix passing of non-exclusive values to randomInteger
add tests for checking the uniformity and bias of the values returned by randomInteger

tomauth0 committed:
exclusive param notes

tomauth0 committed:
cleanup of tests for clearer explanations

luuuis commented on test/randomInteger.tests.js:
```diff
@@ -26,4 +26,85 @@ describe("randomInteger", function () {
+            return Math.pow(num - pdfMean, 2);
+        }));
+
+        expect(pdfVariance).is.lessThan(1.67e-7)
```
Out of curiosity, does this fail against the previous version of the code? :)
